crates/core/src/rlp/structs.rs

```diff
@@ -0,0 +1,75 @@
+use super::{
+    decode::{decode_rlp_item, RLPDecode},
+    error::RLPDecodeError,
+};
+
+/// # Struct decoding helper.
+///
+/// Used to decode a struct from RLP format.
+/// The struct's fields must implement [`RLPDecode`].
+/// The struct is expected as a list, with its values being the fields
+/// in the order they are passed to [`Decoder::decode_field`].
+#[derive(Debug)]
+pub struct Decoder<'a> {
+    payload: &'a [u8],
+    remaining: &'a [u8],
+}
+
+impl<'a> Decoder<'a> {
+    pub fn new(buf: &'a [u8]) -> Result<Self, RLPDecodeError> {
+        match decode_rlp_item(buf)? {
+            (true, payload, remaining) => Ok(Self { payload, remaining }),
+            (false, _, _) => Err(RLPDecodeError::UnexpectedList),
+        }
+    }
+
+    pub fn decode_field<T: RLPDecode>(self, name: &str) -> Result<(T, Self), RLPDecodeError> {
+        let (field, rest) = <T as RLPDecode>::decode_unfinished(self.payload)
+            .map_err(|err| field_decode_error::<T>(name, err))?;
+        let updated_self = Self {
+            payload: rest,
+            ..self
+        };
+        Ok((field, updated_self))
+    }
+
+    pub fn finish(self) -> Result<&'a [u8], RLPDecodeError> {
+        if self.payload.is_empty() {
+            Ok(self.remaining)
+        } else {
+            Err(RLPDecodeError::MalformedData)
+        }
+    }
+}
+
+fn field_decode_error<T>(field_name: &str, err: RLPDecodeError) -> RLPDecodeError {
+    let typ = std::any::type_name::<T>();
+    let err_msg = format!("Error decoding field '{field_name}' of type {typ}: {err}");
+    RLPDecodeError::Custom(err_msg)
+}
+
+#[cfg(test)]
```

Comment:
We can also test a RLPDecode impl for Simple, something like:

```rust

#[cfg(test)]
mod tests {
    use crate::rlp::{decode::RLPDecode, structs::Decoder};

    #[test]
    fn test_decoder_simple_struct() {
        #[derive(Debug, PartialEq, Eq)]
        struct Simple {
            pub a: u8,
            pub b: u16,
        }

        impl RLPDecode for Simple {
            fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), crate::rlp::error::RLPDecodeError> {
                let decoder = Decoder::new(rlp)?;
                let (a, decoder) = decoder.decode_field("a")?;
                let (b, decoder) = decoder.decode_field("b")?;
                let rest = decoder.finish()?;
                Ok((Simple { a, b }, rest))
            }
        }

        let encoded = vec![0xc2, 0x3d, 0x4b];
        let expected = Simple { a: 61, b: 75 };

        let decoded = Simple::decode(&encoded).unwrap();
        assert_eq!(decoded, expected);
    }
}
```

So we have an example about how to use the decoder. 
Also I think we can add the example as part of the Decoder documentation 🤔 .

Reply:
Good idea!